### DISCUSSION TITLE
Extracting CYM planes for printing

### DISCUSSION THREAD

**Discussion by thatgitsam** [ImageMagick/ImageMagick#6428] [Digital Image Processing] [answered]
We are trying to break images down to send to a DyeSub printer via it's API, but cannot get the format of the data correct. It requires 8bit BMP files in greyscale for each plane of CMY, holding only the data for that plane.

We are using `convert sample.jpg -colorspace CMY -separate test.bmp` which produces 3 files which look correct, but appear to have 3x the pixel data that we require.

Using GIMP, if we open one of these generated files and use Image->Mode->Grayscale, the resulting image reduces by 2/3, however we are doing this with scripts and would prefer to keep this in ImageMagick's world. However using `convert test-1.bmp -colorspace GRAY test-1g.bmp` keeps the file at the original file size, and nothing else we have tried seems to do the trick.

Does anyone have any pointers on how to do the equivalent of the GIMP Mode Gray command using ImageMagick?

**Comment by snibgo** [comment]
A simple arithmetical conversion to CMY may not give a good result. It may be better to use an ICC colour profile that is designed for your printer.

Aside from that, perhaps this does what you want:
```
magick toes.png -colorspace CMY -separate -depth 8 -colorspace Gray test.bmp
```

**Reply by thatgitsam** [reply]
We tried those flags in our attempts, but didn't make any difference.

**Comment by urban-warrior** [comment]
BMP does not have a specific grayscale format.  The closest might be to create the colormapped format by adding `-colors 256` to your command-line.

**Reply by thatgitsam** [reply]
Adding `-colors 256` has got the file size down to the expected ballpark, I'll run it through the system and see if that has got it!

**Comment by thatgitsam** [comment]
More info:

What we need to export is "1 pixel needs 1 byte (=8 bit grey scale)" with a result of 2585088 bytes of pixel data (according to the manufacturer). 

Using `convert sample.jpg -colorspace CMY -separate -depth 8 -colors 256 test.bmp` on an image of 1408x1836, we get 3 bitmap files with pixel data size of 4460222 bytes.

As long as we can get that pixel data out, we can attached it to a custom header and it **should** work. But we cannot get that 1 pixel 1 byte...

**Comment by snibgo** [comment]
> with a result of 2585088 ... on an image of 1408x1836

You won't get that with a BMP output, because 1408x1836 = 2585088, and BMP files have headers.

I suggest the format `gray:`, like this:
```
magick sample.jpg -colorspace CMY -separate -depth 8 +adjoin gray:x.bin
```
This will create 3 files, named x-0.bin, x-1.bin and x-2.bin. Each will have no header, and exactly one byte per pixel.

**Comment by thatgitsam** [comment]
Thank you for the input so far.

This has produced the correct pixel count etc, but does not seem to have produced the right image content.

[This is the result](https://drive.google.com/drive/folders/17nDwqx8-B2IcEOtRtyPoTcPxN_GY3Cgx?usp=sharing)

- sample.jpg is the original
- toCMY is produced using `convert sample.jpg -colorspace CMY -separate -depth 8 test.bmp`
- toRaw is produced using `convert sample.jpg -colorspace CMY -separate -depth 8 +adjoin gray:raw.bin`
- RawJoined is produced by simply copying the pixel data directly into place over the Supplied Cyan file from byte 1088, no changes to header etc, which in theory should be the right format to process

The toCMY look like they are probably right - they feel like negatives for printing (to my albeit non-expert eye). The toRaw finished result looks more like greyscale versions, and is flipped on horizontal axis.

**Comment by snibgo** [comment]
Please show the input image.

**Reply by thatgitsam** [reply]
It should be in the root of that linked folder as sample.jpg, I think we've included all stages there.

**Comment by snibgo** [comment]
Ahh, yes, sample.jpg. Found it.

The toRaw conversions look good to me. The binary files are not viewable, so we can convert them to PNG:
```
magick -size 1408x1836 -depth 8 gray:raw-0.bin x0.png
```
Similarly for "1" and "2". x0.png is:

![x0](https://github.com/ImageMagick/ImageMagick/assets/33812368/8c248ff7-15df-45a4-8502-c2ec3be940ad)

The bird faces the correct way, the three images look "negative", and they are slightly different from each other.

**Comment by thatgitsam** [comment]
That looks good, the result is .png, and we need it in .bmp. When trying `magick -size 1408x1836 -depth 8 gray:raw-0.bin x0.bmp` we the result back to 3x the size, so puts us back where we started.

In the link, the RawJoined folder shows result of taking the RAW binary file data and copying it directly over the existing pixels a correctly headed .bmp file (one of the ones in the Supplied folder), so that can be viewed as it is.

This is so close to being right - we need to produce the above "negatives" in .bmp format with 1 byte per pixel.

**Comment by snibgo** [comment]
I am confused about what you want. BMP files have headers. IM won't create BMP files without headers. If you want BMP files, then:
```
magick sample.jpg -colorspace CMY -separate -posterize 256 -depth 8 -compress None x.bmp
```
This gives 3 files, each 2,586,234 bytes. The ".bin" files above have no headers, and are 2,585,088 bytes each.

**Reply by thatgitsam** [reply]
The confusion is understandable, because according to the printer API:
- The header data must be exactly 1088 bytes
- The pixel data must be exactly 2585088 bytes
- We have to manually (using code) alter/pad the header to fit these specs

This command looks very promising, as it seems to just need 58 bytes padded in. I'll run it though the system and see what happens!

**Comment by thatgitsam** [comment, marked as answer]
It worked - with the addition of a -negate. The colour isn't perfect, but that's a problem for another time.

`convert sample.jpg -negate -colorspace CMY -separate -posterize 256 -depth 8 -compress None x.bmp`

![PXL_20230625_180134107](https://github.com/ImageMagick/ImageMagick/assets/29076171/434ebdb7-daa7-4703-abaf-b8f0a6bffa74)

Thank you for your help, it's been invaluable!
